src/pages/index.js

```diff
@@ -63,27 +102,110 @@ function handleImageClick(cardLink, cardCaption) {
   imagePopup.open(cardLink, cardCaption);
 }
 
+function handleAvatarClick() {
+  editAvatarPopup.open();
+}
+
+function handleDeletePopup(cardId, evt) {
+  deleteVerifyPopup.open();
+  cardIdForDeletion = cardId;
+  cardToDelete = evt.target.closest(".card");
+}
+
+function likeToServer(cardId, cardElement) {
+  if (!cardElement.querySelector(".card__liked_active")) {
+    api.removeLikeFromCard(cardId).catch((err) => console.error(err));
+  } else {
+    api.addLikeToCard(cardId).catch((err) => console.error(err));
+  }
+}
+
 function createCard(imageInfo) {
-  const newCard = new Card(imageInfo, "#card-template", handleImageClick);
+  const newCard = new Card(
+    imageInfo,
+    "#card-template",
+    handleImageClick,
+    handleDeletePopup,
+    likeToServer
+  );
   const cardElement = newCard.generateCard();
   return cardElement;
 }
 
+function renderLoading(isLoading, button) {
+  if (isLoading) {
+    buttonText = button.textContent;
+    button.textContent = "Saving...";
+  } else {
+    button.textContent = buttonText;
+  }
+}
+
 function handleAddImage(imageInfo) {
-  const addedCardElement = createCard(imageInfo);
-  cardSection.addItem(addedCardElement, "prepend");
-  addFormPopup.close();
-  addFormPopup.clearForm();
-  formValidators["add"].disableSubmitButton();
+  renderLoading(true, addPopup.querySelector(config.submitButtonSelector));
+  api
+    .addCard(imageInfo)
+    .then((res) => {
+      let addedCardElement = createCard(res);
+      cardSection.addItem(addedCardElement, "prepend");
+    })
+    .catch((err) => console.error(err))
+    .finally(() => {
+      renderLoading(false, addPopup.querySelector(config.submitButtonSelector));
+      addFormPopup.close();
+      addFormPopup.clearForm();
+      formValidators["add"].disableSubmitButton();
```

Comment:
You need to close popups only in blocks `then` after a **successful response** from the server so that when there is an error it would be opened (the inputs will not be cleared and the user could send the data again to the server) and if you change the button text you could see the change as well

Reply:
You need to **reset** the form only  in block `then` not to do that immediately after `submit` because when there is an error the popup will still be opened and the user can try to send the data again to the server without having to type the data into the inputs once more